larreco/TrackFinder/PMAlgTrackMaker_module.cc

```diff
@@ -0,0 +1,1368 @@
+//
+// Progress:
+//    May-June 2015:   track finding and validation, growing tracks by iterative merging of matching
+//                     clusters, no attempts to build multi-track structures, however cosmic tracking
+//                     works fine as they are sets of independent tracks
+//    June-July 2015:  merging track parts within a single tpc and stitching tracks across tpc's
+//    August 2015:     optimization of track-vertex structures (so 3D vertices are also produced)
+//    November 2015:   use track-shower splitting at 2D level, then tag low-E EM cascades in 3D
+//                     note: the splitter is not finished and not as good as we want it
+//    January 2016:    output of track-vertex finding as a tree of PFParticles, refined vertexing
+//                     code, put vertex at front of each track, flip whole track structures to
+//                     selected, direction (beam/down/dQdx), use any pattern reco stored in
+//                     PFParticles as a source of associated clusters
+//    Mar-Apr 2016:    kinks finding, EM shower direction reconstructed for PFPaarticles tagged as
+//                     electrons
+//
+////////////////////////////////////////////////////////////////////////////////////////////////////
+
+#include "art/Framework/Core/EDProducer.h"
+#include "art/Framework/Core/ModuleMacros.h"
+#include "art/Framework/Principal/Event.h"
+#include "art/Framework/Principal/Handle.h"
+#include "art/Framework/Principal/Run.h"
+#include "art/Framework/Principal/SubRun.h"
+#include "art/Framework/Services/Optional/TFileService.h"
+#include "art/Framework/Services/Optional/TFileDirectory.h"
+#include "art/Utilities/InputTag.h"
+#include "fhiclcpp/ParameterSet.h"
+#include "messagefacility/MessageLogger/MessageLogger.h"
+
+// LArSoft includes
+#include "larcore/Geometry/Geometry.h"
+#include "larcore/Geometry/TPCGeo.h"
+#include "larcore/Geometry/PlaneGeo.h"
+#include "larcore/Geometry/WireGeo.h"
+#include "lardata/RecoBase/Hit.h"
+#include "lardata/RecoBase/Cluster.h"
+#include "lardata/RecoBase/PFParticle.h"
+#include "lardata/RecoBase/Track.h"
+#include "lardata/RecoBase/TrackHitMeta.h"
+#include "lardata/RecoBase/Vertex.h"
+#include "lardata/RecoBase/SpacePoint.h"
+#include "lardata/AnalysisBase/T0.h" 
+#include "lardata/DetectorInfoServices/DetectorPropertiesService.h"
+#include "lardata/Utilities/AssociationUtil.h"
+
+#include "larreco/RecoAlg/ProjectionMatchingAlg.h"
+#include "larreco/RecoAlg/PMAlgVertexing.h"
+#include "larreco/RecoAlg/PMAlg/Utilities.h"
+#include "larreco/RecoAlg/PMAlg/PmaTrkCandidate.h"
+
+#include "TTree.h"
+#include "TMath.h"
+
+#include <memory>
+
+namespace trkf {
+
+typedef std::map< size_t, std::vector<double> > dedx_map;
+typedef std::map< unsigned int, std::vector< art::Ptr<recob::Hit> > > view_hitmap;
+typedef std::map< unsigned int, view_hitmap > tpc_view_hitmap;
+typedef std::map< unsigned int, tpc_view_hitmap > cryo_tpc_view_hitmap;
```

Comment:
Suggest that the experts comment on the design of this data structure.  It's essentially a three index container of std::vector<art::Ptr<recob::Hit>; the three indices are IDs for a cryostat, a TPC within a cryostat and a view within a  TPC.   All three indices are always small (or smallish) integers. 


1. Is std::map appropriate for the outer three containers?   
2. Is the right answer to invent a dense integer ID for the (cryo,tpc,view) tuple and then use a std::array as the outer container?  
3.  If so, the dense ID could be defined here but it would be better if it were defined in the geometry subsystem; this involves cooperation across multiple experiments - is it practical?
4.  In this data structure, do all Ptr's always point into a single data product - or is it possible that they may point into more than one data product?  If the former then switch to bare  pointers and keep track of the product ID separately. This will have both speed and memory footprint improvements.
5. If all Ptrs point to a single data product, does it make sense to design that data product to support the 3-index structure used here?  Could other algorithms use such a reorganization?  Work with other algorithm developers to understand how to structure the raw hit containers to support the algorithms.
6. The code would be easier for next-generation maintainer/developer to follow if you have types for the cryo ID, tic ID and view ID.  Just a typedef helps; doing the work get develop an opaque integer will turn crossed-indexing into a compile time error.